### PR TITLE
feat(groq): GitHub Action that automatically adds a specified label to the pull request that triggered the workflow.

### PR DESCRIPTION
**Pull request by polsala**
## Implementation Summary
- **Utility**: nightly-pr-labeler
- **Provider**: groq
- **Location**: `github-actions/nightly-nightly-pr-labeler-8`
- **Files Created**: 4
- **Description**: GitHub Action that automatically adds a specified label to the pull request that triggered the workflow.

## Rationale
- Automated proposal from the Groq generator delivering a fresh community utility.
- This utility was generated using the **groq** AI provider.

## Why safe to merge
- Utility is isolated to `github-actions/nightly-nightly-pr-labeler-8`.
- README + tests ship together (see folder contents).
- No secrets or credentials touched.
- All changes are additive and self-contained.

## Test Plan
- Follow the instructions in the generated README at `github-actions/nightly-nightly-pr-labeler-8/README.md`
- Run tests located in `github-actions/nightly-nightly-pr-labeler-8/tests/`

## Links
- Generated docs and examples committed alongside this change.

## Mock Justification
- Not applicable; generator did not introduce new mocks.

### CODE REVIEW

**Comment by polsala** [comment]
## 🤖 Review by GROQ Agent

## ✅ What's solid
- **Self‑contained composite action** – all files live under `github-actions/nightly-nightly-pr-labeler-8`, keeping the utility isolated from the rest of the repo.  
- **Clear Bash implementation** – the script validates required env vars, extracts the PR number with `jq`, builds a minimal JSON payload, and performs a single `curl` request.  
- **Fail‑fast behavior** – `set -euo pipefail` ensures any unexpected condition aborts the action early, preventing silent failures.  
- **Straight‑forward inputs** – the `action.yml` surface only the two needed parameters (`github-token` and `label`), making the action easy to adopt.  
- **License & attribution** – MIT header in the README makes reuse unambiguous.  

## 🧪 Tests
- **Good start with a Bash‑only test** that mocks `curl` and validates the exact request that would be sent.  
- **Actionable improvements**
  - **Cleanup temporary files** – the test creates a temporary `GITHUB_EVENT_PATH` but never removes it. Add a `trap` to delete the file after the test runs:
    ```bash
    export GITHUB_EVENT_PATH=$(mktemp)
    trap 'rm -f "$GITHUB_EVENT_PATH"' EXIT
    ```
  - **Test error paths** – currently only the happy path is covered. Add cases for:
    - Missing environment variables (e.g., `LABEL` unset) → script should exit with a non‑zero status and a helpful message.
    - Event payload without a `pull_request.number` (e.g., a `push` event) → verify the “No pull_request number found” branch.
  - **Use a test framework** – consider BATS‑core or shunit2 to get richer assertions (`assert_success`, `assert_output_contains`, etc.) and better reporting.
  - **Avoid alias side‑effects** – the `alias curl=mock_curl` persists for the remainder of the shell session. Prefer overriding via `PATH`:
    ```bash
    mkdir -p "$TMPDIR/mock-bin"
    echo -e '#!/usr/bin/env bash\nmock_curl "$@"' > "$TMPDIR/mock-bin/curl"
    chmod +x "$TMPDIR/mock-bin/curl"
    PATH="$TMPDIR/mock-bin:$PATH" bash .../label_pr.sh
    ```
    This isolates the mock to the test process only.

## 🔒 Security
- **Token handling** – the script passes the token via an `Authorization: token $TOKEN` header, which is correct. Ensure the token never appears in logs:
  - Do **not** add `set -x` or any debugging flag that would echo the header.
  - The final `echo "Label '$LABEL' added …"` is safe because it does not include the token.
- **Input sanitisation** – the label value is interpolated directly into a JSON payload via `jq`. This is safe because `jq` handles proper escaping, but you might want to explicitly validate that the label conforms to GitHub’s label naming rules (max 50 characters, no control characters). A quick guard:
  ```bash
  if [[ ${#LABEL} -gt 50 || "$LABEL" =~ [[:cntrl:]] ]]; then
    echo "Invalid label name"
    exit 1
  fi
  ```
- **Dependency on `jq`** – the script assumes `jq` is present in the runner image. Document this requirement in the README (or add a step that installs it) to avoid runtime failures that could be mis‑interpreted as permission issues.

## 🧩 Docs / Developer Experience
- **README usage path** – the example uses `uses: ./utils/github-actions/nightly-pr-labeler`, but the actual location is `github-actions/nightly-nightly-pr-labeler-8`. Update the snippet to reflect the correct relative path or publish the action under a proper namespace.
- **Prerequisites section** – add a short note that the runner must have `jq` installed (most GitHub‑hosted runners do, but it’s good to be explicit).
- **Versioning guidance** – since the folder name includes a numeric suffix (`nightly-...-8`), explain how consumers should pin a version (e.g., using a tag or a commit SHA) to avoid accidental breaking changes.
- **Error messages** – the script prints generic “Missing required environment variables.” Consider echoing which variable is missing to aid debugging:
  ```bash
  missing=()
  [[ -z "$EVENT_PATH" ]] && missing+=("GITHUB_EVENT_PATH")
  [[ -z "$REPO" ]] && missing+=("GITHUB_REPOSITORY")
  [[ -z "$TOKEN" ]] && missing+=("GITHUB_TOKEN")
  [[ -z "$LABEL" ]] && missing+=("LABEL")
  if (( ${#missing[@]} )); then
    echo "Missing required environment variables: ${missing[*]}"
    exit 1
  fi
  ```
- **Link to GitHub API docs** – a small “See the GitHub REST API docs for adding labels” line would help newcomers understand the underlying request.

## 🧱 Mocks / Fakes
- **Current mock approach works** – redefining `curl` via an alias captures the request details without network traffic.  
- **Potential fragility** – if the script under test ever invokes `curl` with additional flags (e.g., `--fail`), the mock’s argument parser may need updates. Using a wrapper script placed earlier in `PATH` (as shown in the Tests section) makes the mock more resilient to flag changes.  
- **Isolation** – ensure the mock does not leak into other tests or CI steps. Wrap the mock setup/teardown inside the test file or a dedicated test harness.  

---

Overall, the action is well‑structured and functional. The suggestions above focus on tightening the test suite, clarifying documentation, and hardening security‑related edge cases, which will make the utility more robust for downstream consumers.

**Comment by polsala** [comment]
## 🤖 Review by GEMINI Agent

✅ What's solid
*   The action's purpose is clear, and its inputs (`github-token`, `label`) are well-defined and appropriately marked as required in `action.yml` and `README.md`.
*   The use of a `composite` action is suitable for encapsulating a shell script, providing a clean interface for users.
*   The `label_pr.sh` script demonstrates good shell scripting practices, including `set -euo pipefail` for robustness and explicit checks for missing environment variables.
*   The script correctly extracts the PR number using `jq` and constructs the GitHub API request with the appropriate `POST` method, authorization header, and JSON payload.
*   The `README.md` provides comprehensive documentation, including a clear overview, input descriptions, an explanation of how the action works, and a practical usage example.

🧪 Tests
*   The `test_label_pr.sh` script employs an effective strategy by mocking `curl` to prevent actual network requests, allowing for isolated and fast unit testing of the script's API interaction logic.
*   The `mock_curl` function successfully captures the HTTP method, URL, and data payload, enabling precise assertions about the generated API call.
*   The test correctly sets up a temporary `GITHUB_EVENT_PATH` with a mock JSON payload, ensuring the script receives expected input during testing.

*Actionable feedback:*
*   Consider adding a test case to verify the script's behavior when `jq` fails to extract a `pull_request.number` (e.g., if the event payload is not a pull request event or is malformed). This would confirm that the script correctly exits with an error message as intended by the `if [[ "$PR_NUMBER" == "null" || -z "$PR_NUMBER" ]]; then ... exit 1; fi` block.
*   While the primary focus is on the `curl` command, the test could optionally assert the final `echo` output of the `label_pr.sh` script for full coverage of its console interactions.

🔒 Security
*   The action correctly utilizes `secrets.GITHUB_TOKEN` for authentication, which is the recommended and secure method for GitHub Actions.
*   Passing the token as an environment variable to the script is secure within the action's execution context.
*   The `set -euo pipefail` command in the shell script significantly enhances security by ensuring the script exits immediately upon encountering an error or an unset variable, preventing unexpected behavior.
*   The use of `jq` for JSON parsing and construction is generally more secure than manual string manipulation, reducing the risk of injection vulnerabilities.

*Actionable feedback:*
*   The `github-token` input description in `action.yml` and `README.md` currently specifies `repo` scope. For adding labels, the more granular `issues:write` or `pull-requests:write` permission is sufficient. Updating the description to reflect the principle of least privilege would be beneficial, even though `secrets.GITHUB_TOKEN` often has broader default permissions.
    ```diff
    --- a/github-actions/nightly-nightly-pr-labeler-8/README.md
    +++ b/github-actions/nightly-nightly-pr-labeler-8/README.md
    @@ -10,7 +10,7 @@
     ## Inputs
    
     | Name | Description | Required |
    -|------|-------------|----------|
    -| `github-token` | A GitHub token with `repo` scope (usually `secrets.GITHUB_TOKEN`). | Yes |
    +|------|-------------|----------|
    +| `github-token` | A GitHub token with `issues:write` or `pull-requests:write` permission (usually `secrets.GITHUB_TOKEN`). | Yes |
     | `label` | The label to add to the PR. | Yes |
    ```
*   The `curl` command in `label_pr.sh` does not explicitly handle potential API errors (e.g., 404 Not Found, 422 Unprocessable Entity if the label doesn't exist or permissions are insufficient). While `set -e` would catch a `curl` command failure, it might not interpret a successful HTTP response with an API error message as a failure. Consider adding a check for the API response to provide more informative error messages.
    ```bash
    # ... (previous script content) ...
    
    RESPONSE=$(curl -s -X POST \
      -H "Authorization: token $TOKEN" \
      -H "Accept: application/vnd.github+json" \
      "$API_URL" \
      -d "$PAYLOAD")
    
    # Check if the response contains an error message from the GitHub API
    if echo "$RESPONSE" | jq -e '.message' > /dev/null; then
      echo "Error adding label: $(echo "$RESPONSE" | jq -r .message)"
      exit 1
    fi
    
    echo "Label '$LABEL' added to PR #$PR_NUMBER."
    ```

🧩 Docs/DX
*   The `README.md` is well-structured and provides all necessary information for users to understand, configure, and use the action effectively.
*   The usage example is clear and directly demonstrates how to integrate the action into a workflow.
*   The `action.yml` includes a descriptive `name` and `description` for the action, along with clear input descriptions, which enhances its usability and discoverability.

*Actionable feedback:*
*   The usage example in `README.md` currently uses `uses: ./utils/github-actions/nightly-pr-labeler`. Based on the PR's file paths, the action will be located at `github-actions/nightly-nightly-pr-labeler-8`. The example should be updated to reflect the correct relative path to avoid user confusion.
    ```diff
    --- a/github-actions/nightly-nightly-pr-labeler-8/README.md
    +++ b/github-actions/nightly-nightly-pr-labeler-8/README.md
    @@ -29,7 +29,7 @@
     jobs:
       label:
         runs-on: ubuntu-latest
         steps:
           - name: Label PR
    -        uses: ./utils/github-actions/nightly-pr-labeler
    +        uses: ./github-actions/nightly-nightly-pr-labeler-8
             with:
               github-token: ${{ secrets.GITHUB_TOKEN }}
               label: "automated"
    ```
*   Consider adding a small note in the `README.md` about the prerequisite that the specified `label` must already exist in the repository, as the action will fail if it attempts to add a non-existent label.

🧱 Mocks/Fakes
*   The `mock_curl` function is an excellent example of how to effectively mock external commands in bash, allowing for isolated testing of the `label_pr.sh` script's core logic.
*   The mock accurately captures the essential components of the `curl` command (method, URL, data), which are critical for verifying the correctness of the API request.
*   The use of `mktemp` to create a temporary file for `GITHUB_EVENT_PATH` is a robust way to provide fake input data for testing, ensuring test isolation and cleanup.

*Actionable feedback:*
*   The current `mock_curl` implementation ignores HTTP headers (`-H`). While sufficient for the current test, if future requirements involve asserting specific headers (e.g., `Accept` header versions), the mock would need to be extended to capture and verify these as well.
*   To complement the security feedback, the mock could be enhanced to simulate different HTTP response codes or error payloads from the GitHub API. This would allow for more comprehensive testing of the `label_pr.sh` script's error handling, should it be implemented.
